File: distance/edit_distance.py
class EditDistance():
    x = None
    y = None
    match = None
    mismatch = None
    matrix = []
    shift = None
    score = None

    def __init__(self, x, y):
        self.x = x
        self.y = y
        self._calculate_matrix()
        self.score = self.matrix[-1][-1]
    
    def __str__(self):
        return self.score

    def _calculate_matrix(self):
        x = " " + self.x
        y = " " + self.y
        matrix = [ [0] + [self.shift * i for i in range(1, len(y))]]
        for i in range(1, len(x)):
            row = [self.shift * i]
            for j in range(1, len(y)):
                sij = self.match if x[i] == y[j] else self.mismatch
                row.append(min(
                    matrix[i-1][j-1] + sij,
                    matrix[i-1][j] + self.shift,
                    row[j-1] + self.shift
                ))
            matrix.append(row)
        self.matrix = matrix

class HammingDistance(EditDistance):
    match = 0
    mismatch = 1
    shift = float("inf")

    def __init__(self, x, y):
        if len(x) != len(y):
            raise ValueError("Undefined for sequences of unequal length")
        super().__init__(x, y)

class LevenshteinDistance(EditDistance):
    match = 0
    mismatch = 1
    shift = 1

File: cpp/distance/EditDistance.cpp
#include "EditDistance.h"
#include <algorithm>
#include <iostream>


EditDistance::EditDistance(const std::string &x, const std::string &y, double shift, double match, double mismatch) :  
    x_(" " + x), y_(" " + y), shift_(shift), match_(match), mismatch_(mismatch)
{
    matrix_ = new std::vector<std::vector<double> >;
    computeMatrix();
}

void EditDistance::computeMatrix() {
    std::vector<double> row;

    for (int i = 0; i < int(x_.length()); ++i) {
        std::vector<double> row;
        for (int j = 0; j < int(y_.length()); ++j) {
            if(i == 0) {
                row.push_back(j == 0 ? 0 : shift_ * j);
            } else {
                if(j == 0) {
                    row.push_back(shift_ * i);
                    continue;
                }
                double sij = x_[i] == y_[j] ? match_ : mismatch_;
                double new_sij = std::min((*matrix_)[i-1][j-1] + sij, (*matrix_)[i-1][j] + shift_);
                new_sij = std::min(new_sij, row[j-1] + shift_);
                row.push_back(new_sij);
            }
        }
        matrix_->push_back(row);
    }
    score_ = (*matrix_)[x_.length()-1][y_.length()-1];
}

void EditDistance::printMatrix() {
    using namespace std;
    for (int i = 0; i < matrix_->size(); ++i) {
        for (int j = 0; j < (*matrix_)[i].size(); ++j) {
            if(j == 0) cout << (i == 0 ? "[[" : " [");
            cout << (*matrix_)[i][j];
            if(j < (*matrix_)[i].size()-1) cout << ", "; 
            else if(i < matrix_->size()-1) cout << "],\n";
            else cout << "]]\n";
        }
    }
}
